src/Coordination/KeeperServer.cpp

```diff
@@ -619,6 +613,82 @@ nuraft::cb_func::ReturnCode KeeperServer::callbackFunc(nuraft::cb_func::Type typ
         }
     }
 
+    if (!keeper_context->local_logs_preprocessed)
+    {
+        const auto preprocess_logs = [&]
+        {
+            auto log_store = state_manager->load_log_store();
+            if (last_log_idx_on_disk > 0 && last_log_idx_on_disk > state_machine->last_commit_index())
+            {
+                auto log_entries = log_store->log_entries(state_machine->last_commit_index() + 1, last_log_idx_on_disk + 1);
+
+                size_t preprocessed = 0;
+                LOG_INFO(log, "Preprocessing {} log entries", log_entries->size());
+                auto idx = state_machine->last_commit_index() + 1;
+                for (const auto & entry : *log_entries)
+                {
+                    if (entry && entry->get_val_type() == nuraft::log_val_type::app_log)
+                        state_machine->pre_commit(idx, entry->get_buf());
+
+                    ++idx;
+                    ++preprocessed;
+
+                    if (preprocessed % 50000 == 0)
+                        LOG_TRACE(log, "Preprocessed {}/{} entries", preprocessed, log_entries->size());
+                }
+                LOG_INFO(log, "Preprocessing done");
+            }
+            else
+            {
+                LOG_INFO(log, "All local log entries preprocessed");
+            }
+            keeper_context->local_logs_preprocessed = true;
+        };
+
+        switch (type)
+        {
+            case nuraft::cb_func::InitialBatchCommited:
+            {
+                preprocess_logs();
+                break;
+            }
+            case nuraft::cb_func::GotAppendEntryReqFromLeader:
+            {
+                auto & req = *static_cast<nuraft::req_msg *>(param->ctx);
+
+                if (req.get_commit_idx() == 0 || req.log_entries().empty())
+                    break;
+
+                auto last_committed_index = state_machine->last_commit_index();
+                // Actual log number.
+                auto index_to_commit = std::min({last_log_idx_on_disk, req.get_last_log_idx(), req.get_commit_idx()});
+
+                if (index_to_commit > last_committed_index)
+                {
+                    LOG_TRACE(log, "Trying to commit local log entries, committing upto {}", index_to_commit);
+                    raft_instance->commitLogs(index_to_commit, true);
+                    if (!keeper_context->local_logs_preprocessed)
```

Comment:
It fills like this if should be above previous line

Reply:
We manually commit in this thread all the local logs which we can. All the other logs should be preprocessed inside the callback during the commit. This basically asserts that we don't have any local log uncommitted or unprocessed.